scripts/migration/courses-populate.ts

```diff
@@ -0,0 +1,171 @@
+/* eslint-disable no-console */
+/* eslint-disable no-await-in-loop */
+/**
+ * This script retrieves data from all courses from all available
+ * class rosters and pushes them to the 'courses' collection in
+ * Firebase.
+ *
+ * This script also creates a mapping from courses to the rosters
+ * that the course is available in.
+ *
+ *
+ * This script also creates a mapping from crseId to its
+ * catalogNbr.
+ *
+ *  Motivation for this script as well as the schema
+ * for the 'courses' collection can be found here:
+ * https://tinyurl.com/5n86j4p7
+ */
+
+import fetch from 'node-fetch';
+import { FieldValue } from 'firebase-admin/firestore';
+import {
+  coursesCollection,
+  availableRostersForCourseCollection,
+  crseIdToCatalogNbrCollection,
+} from '../firebase-config';
+import { CourseFullDetail } from '../../src/requirements/types';
+
+/** A helper function to generate a wait promise. Used for cooldown to limit API usage. */
+const wait = (time: number) =>
+  new Promise<void>(resolve => {
+    setTimeout(() => resolve(), time);
+  });
+
+/* Retrieves the rosters available on the class roster API */
+const retrieveAvailableRosters = async () => {
+  const rosters: readonly string[] = await fetch(
+    'https://classes.cornell.edu/api/2.0/config/rosters.json'
+  )
+    .then(res => res.json())
+    .then(jsonRes => jsonRes.data.rosters)
+    .then(jsonRosters => jsonRosters.map(jsonRoster => jsonRoster.slug));
+
+  return rosters;
+};
+
+/* Retrieves the subjects across for a roster and creates a list of {roster: string, subject: string} objects */
+const retrieveAvailableSubjects = async (roster: string) => {
+  const subjects: readonly { roster: string; subject: string }[] = await fetch(
+    `https://classes.cornell.edu/api/2.0/config/subjects.json?roster=${roster}`
+  )
+    .then(res => res.json())
+    .then(jsonRes => jsonRes.data.subjects)
+    .then(jsonSubjects =>
+      jsonSubjects.map(jsonSubject => ({ roster, subject: jsonSubject.value }))
+    );
+  return subjects;
+};
+
+/* Retrieves and formats available courses for a {roster: string, subject: string} object */
+const retrieveAvailableCourses = async (roster: string, subject: string) => {
+  const courses: CourseFullDetail[] = await fetch(
+    `https://classes.cornell.edu/api/2.0/search/classes.json?roster=${roster}&subject=${subject}`
+  )
+    .then(res => res.json())
+    .then(jsonRes => jsonRes.data.classes)
+    .then(jsonCourses => jsonCourses.map(jsonCourse => courseFieldFilter(jsonCourse)));
+  return courses;
+};
+
+/* Helper functions to clean and format class roster data */
+const cleanField = (value: string | null | undefined) =>
+  value?.replace(/\u00a0/g, ' ') || undefined;
```

Comment:
iirc @noschiff already made something like this

Reply:
Yeah he did, I looked into his implementation and did not see any export for the his `cleanField`function so I was hesitant to use it. If you guys are cool with it, I can go ahead and export it so I can use it here?

Reply:
Yeah, or maybe pull in out into a utilities file?

Reply:
Yeah maybe pull it out into a utilities file and then import it from both scripts?

Reply:
It's honestly not a good function at all. I wrote it in #720 just because it was a marginal improvement over what we had, and it was not the purpose of the PR.

Reply:
@zachary0kent also… this _seems_ to work fine because of the `/g`, but is it bad practice to use a global regex without replaceAll()?

Reply:
I will use it for now, and we can open another task to make a better `cleanField` function. This PR is already getting pretty big.